src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesVsPatterns.cs

```diff
@@ -0,0 +1,1838 @@
+}
+class Program
+{
+    void M(Base? b)
+    {
+        if (b is Derived { Value: null })
+            b.Value.ToString();
+    }
+}";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void RecursivePatternNullInferenceWithDowncast_02()
+        {
+            var source = @"
+class Base
+{
+    public object Value = """";
+}
+class Derived : Base
+{
+}
+class Program
+{
+    void M(Base? b)
+    {
+        if (b is Derived { Value: null })
+            b.Value.ToString(); // 1
+    }
+}";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (14,13): warning CS8602: Possible dereference of a null reference.
+                //             b.Value.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b.Value").WithLocation(14, 13));
+        }
+
+        [Fact]
+        public void TuplePatternNullInference_01()
+        {
+            var source = @"
+class Program
+{
+    void M((object, object) t)
+    {
+        if (t is (1, null))
+        {
+        }
+        else
+        {
+            t.Item2.ToString(); // 1
+            t.Item1.ToString();
+        }
+    }
+}";
+            var comp = CreateNullableCompilation(source);
+            comp.VerifyDiagnostics(
+                // (11,13): warning CS8602: Possible dereference of a null reference.
+                //             t.Item2.ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "t.Item2").WithLocation(11, 13));
+        }
+
+        [Fact]
+        public void MultiplePathsThroughDecisionDag_01()
+        {
+            var source = @"
+class Program
+{
+    bool M1(object? o, bool cond = true)
+    {
+        o = 1;
+        switch (o)
+        {
+            case null:
+                throw null!;
+            case """" when M1(o = null):
+                break;
+            default:
+                if (cond) o.ToString(); // 1
+                break;
+        }
+
+        return cond;
+    }
+    bool M2(object? o, bool cond = true)
+    {
+        o = 1;
+        switch (o)
+        {
+            case """" when M2(o = null):
+                break;
+            default:
+                if (cond) o.ToString(); // 2
+                break;
+        }
+
+        return cond;
+    }
+    bool M3(object? o, bool cond = true)
+    {
+        o = 1;
+        switch (o)
+        {
+            case null:
+                throw null!;
+            default:
+                if (cond) o.ToString();
+                break;
+        }
+
+        return cond;
+    }
+    bool M4(object? o, bool cond = true)
+    {
+        o = 1;
+        switch (o)
+        {
+            case null:
+                throw null!;
+            case """" when M4(o = null):
+                break;
+            case var q:
+                q.ToString(); // 3 (!?)
```

Comment:
>(!?) [](start = 35, length = 4)

Is there a bug for this?

Reply:
It is not a bug.  It is because we specified that we should act as if switching on the original variable.  The assignment `o = null` can reach here, so `q` is considered possibly null when it captures `o` on line 1183.  At runtime, we actually copy the input variable `o` at entry, so no null can actually reach here.  This is a slightly counterintuitive result of our decision to act as if working with the original value rather than a copy.

---
In reply to: [269678277](https://github.com/dotnet/roslyn/pull/34273#discussion_r269678277) [](ancestors = 269678277)